docs/synapse/devguide.rst

```diff
@@ -0,0 +1,331 @@
+                            'doc': 'A custom event ID from some other system.'}),
+
+                        ('x:foo:bar', ('int', {'min': 100, 'max': 10000}), {
+                            'doc': 'A custom integer property with a fixed range.'}),
+                    ),
+
+                    'forms': (
+
+                        # declare a new node type
+
+                        ('x:foo:event', {}, (
+
+                            # declare secondary properties
+                            ('time', ('time', {}), {
+                                'doc': 'The time of the custom event.'}),
+
+                            ('ipv4', ('inet:ipv4', {}), {
+                                'doc': 'The ipv4 associated with the custom event.'}),
+
+                            ('bar', ('x:foo:bar', {}), {
+                                'doc': 'The custom bar property associated with the custom event.'}),
+                        )),
+                    ),
+
+                }),
+            )
+
+In the above example, we can see that the model extension implements a custom form ``x:foo:event`` with properties using a mix of existing types and custom types.  By using existing types where possible, the model extension easily integrates with the existing data model, allowing seamless pivots to and from the custom nodes.  Additionally, custom field specific to the deployment environment allow knowledge within the Cortex to be linked to external records.::
+
+    cli> storm x:foo:event:time*range=(20190505, 20190506) -> inet:ipv4
+
+The Storm query above returns all the ``inet:ipv4`` nodes associated with events on 2019/05/05.
+
+As the mainline data model grows, so does the power of analysis using Synapse.  For data model extensions that are (or could be) generalized, we strongly encourage analysts and developers to discuss their data model ideas with Vertex for potential inclusion in the mainline model.  Check out the slack chat or email info@vertex.link for more info.
+
+Custom Storm Commands
+---------------------
+
+A ``CoreModule`` subclass may extend the commands available in the Storm query language implementing the ``getStormCmds`` API to return ``synapse.lib.storm.Cmd`` subclasses.
+
+WARNING: It is extremely important that developers are familiar with the Python 3.7 asyncio programming paradigm when implementing ``Cmd`` subclasses!  Any blocking API calls made from within a ``Cmd`` extension will block *all* execution for the entire Cortex and Telepath multiplexor!
+
+A ``Cmd`` subclass implements several methods to facilitate both parsing and execution.  The Storm query runtime is essentially a pipeline of nodes being lifted, pivoted, and filtered.  The main execution of a ``Cmd`` subclass is handled by the ``execStormCmd`` method which iterates over the nodes it is given and yield nodes which are the result.  This architecture allows Storm to chain commands and operations together in any way the analyst sees fit.::
```

Comment:
we should point out that we're iterating over the `Node, Path` object tuples.

Reply:
I'll add something.  I've got it in the code example, and wanted to avoid getting too wordy ( by example rather than prose ) but it probably needs a note here.